### DISCUSSION TITLE
A plugin no longer works in 3.3.0 - Can anyone advise or help as to why?

### DISCUSSION THREAD

**Discussion by tegitting** [nicotine-plus/nicotine-plus#2868] [Q&A] [answered]
Nicotine+ version:  3.3.0
Operating System/Distribution: Debian

### Description
The plugin no longer works but now instead throws the following response:

2024-02-08 12:58:55 Plugin autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/usr/lib/python3/dist-packages/pynicotine/pluginsystem.py", line 943, in _trigger_event
    return_value = getattr(plugin, function_name)(*args)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wsl/.local/share/nicotine/plugins/autobahn/__init__.py", line 50, in upload_queued_notification
    self.frame.np.userbrowse.browse_user(user)
    ^^^^^^^^^^
2024-02-08 12:58:55 Plugin autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/usr/lib/python3/dist-packages/pynicotine/pluginsystem.py", line 943, in _trigger_event
    return_value = getattr(plugin, function_name)(*args)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wsl/.local/share/nicotine/plugins/autobahn/__init__.py", line 66, in user_stats_notification
    total_size_text = self.frame.userbrowse.pages[user].share_size_label.get_text()
                      ^^^^^^^^^^

### Expected behavior

The plugin browses the shares of the person downloading from me and checks amount of shares they have. If 0, or lower than a configured value then they are banned from downloading from me.

### Steps to reproduce the bug
I guess you might need to install the addon and try to use it while a person who is not sharing downloads from you.
Unfortunately I cannot find a link to the addon and am not able to attach the .py file - but it worked until i upgraded Nicotine and I hoped that someone may be able to explain the error response.

**Comment by mathiascode** [comment, marked as answer]
The GUI (i.e. self.frame) is no longer accessible from plugins in 3.3.0, since the core no longer knows anything about the GUI. You (or whoever wrote the plugin) will have to take a look at the latest changes in the Leech Detector plugin, and make changes accordingly: https://github.com/nicotine-plus/nicotine-plus/blob/master/pynicotine/plugins/leech_detector/__init__.py

In this case, you no longer need to read the share size from the GUI (`share_size_label`). Simply requesting a user's shares (`self.core.userbrowse.request_user_shares(user)`) will trigger a `user_stats_notification` with stats from the response.

**Reply by tegitting** [reply]
Thank you very much for taking the time to give a details response. I appreciate it.

Kind Regards,

**Reply by iioniis** [reply]
Were you able to get it to work?

**Reply by tegitting** [reply]
Hi, yes I think I was able to make it work using the provided advice but need some more time to test and confirm.
I don't know when I will have time to do that. Did you get this working or do you need it it too?

**Comment by iioniis** [comment]
Here is this:

https://pastebin.com/KQ3yVgcn

**Reply by tegitting** [reply]
The code in the pastebin link will not work due to the frame references as detailed by mathiascode

**Comment by iioniis** [comment]
I have not been able to.  I am very, very green and know almost absolutely nothing about computer languages.  I was actually hoping to get some help in at least making sure I have things in the right directories and things are right to make it will run on my Mac.

**Reply by tegitting** [reply]
Hi, I am not au fait with the MAC file structure but on Linux (ubuntu) I placed the addon here and it was picked up by the client. [username] is where the logged in computer username appears.

/home/[username]/.local/share/nicotine/plugins/autoban/

file is named [underscore][underscore]init[underscore][underscore].py and I made it executable (I think that was required). (Replace [underscore] with actual underscores, for some reason this site was stripping those out.

By that you should at least be able to tell if the client is picking it up and recognises it and if you manage to get that working I can share the modified version I have when I confirm it does the job correctly.

I hope that a future version of the nicotine client can ship with something similar to this in future as I think it could help improve the SLSK network.

**Reply by slook** [reply]
> I made it executable (I think that was required)

The executable attribute shouldn't be required, since the nicotine executable loads the `__init__.py` plugin script as a module.

**Reply by iioniis** [reply]
So put this file: https://github.com/nicotine-plus/nicotine-plus/blob/defdc0e8f071230c3f6ee7fbd239321b755c0699/pynicotine/__init__.py#L4  in the directory specified.  Loaded the client and not sure if it's working.  The plugin debugger is giving me 'user_[action]_[action](user=[user], ip_address=[ip addrress], port=[port], country=[country]'...but all country results for users is [no country].  Is this just a Nicotine process or is this plugin running?  the .py file is in the folder 'autoban' in the said directory that tegitting provided above.  sorry for the noob questions.

**Reply by slook** [reply]
No @iioniis that's the wrong `__init__.py` file is the `pynicotine` core process which has nothing to do with custom plugins.

A plugin that you put in the `plugins` directory should be listed in the Preferences > Plugins dialog (close and reopen the Preferences dialog if needed). If it doesn't appear, then it's in the wrong directory or it's not a valid plugin.

**Reply by tegitting** [reply]
Sorry for the misunderstanding but what I meant was, for you put the file that you linked to on paste bin [https://pastebin.com/KQ3yVgcn] in the folder that I mentioned. This pastebin file should also have the name as specified previously. You will see it show up in the list of plugins and you can configure the setttings for it but as of 3.3.0 it will throw errors *due to referencing frame as pointed out by Mathias* but this was just to make sure you had the files all setup correctly. Once a modified version was ready for 3.3.0 the idea was for you to simply replace the outdated file with the newer one and you could once again be working.

**Reply by iioniis** [reply]
Ohhh okay.  So I just put the pastebin file in directory: /home/[username]/.local/share/nicotine/plugins/autobahn/[pastebin file]

I spelled Autobahn with an 'h' as that is how it is described in the file.  I don't know if that matters or not.  I renamed the pastebin file: '__init__.py'  correct?  I just tried that and it doesn't show up.

**Reply by tegitting** [reply]
For me it had to be named `__init__.py` for it to show up.

**Reply by tegitting** [reply]
Did the directory `/home/[username]/.local/share/nicotine/plugins/`  (where [username] is your own actual computer login name) already exist for you or did you have to create this?

**Comment by iioniis** [comment]
This already existed

**Reply by iioniis** [reply]
But directory as it shows up on my Mac is: /Users/[username]/.local/share/nicotine/plugins     with my hard drive root before 'users'.

**Reply by tegitting** [reply]
Ah, ok, and the file is named `__init__.py` ? (thats 2 underscores each side of init).

**Comment by iioniis** [comment]
Yes.

**Comment by tegitting** [comment]
Sorry about that then, I have no idea why the client would not be recognising that. Hopefully someone familiar with MAC can say what is needed to be done to get the client to pickup the plugin. 

Once I had did the above, I was able to configure the plugin via the normal menus but then noticed as of 3.3.0 it was throwing errors back at me which is what started this whole thread.

**Comment by slook** [comment]
It is needed to put the 2 files in a sub-folder such as for example:

`/Users/[username]/.local/share/nicotine/plugins/autobahn/PLUGININFO`
`/Users/[username]/.local/share/nicotine/plugins/autobahn/__init__.py`

If either one of those files is missing then it won't work.

**Reply by iioniis** [reply]
I'm sure this is not the correct file but I put 'PLUGININFO' from https://github.com/ithcy/nicotine-plugins/blob/c8977b44c814937ea01de4ace5de43c93a95e7cb/Autobahn/PLUGININFO  in the sub-folder with '__init__.py'

This is not the correct file right?

**Reply by slook** [reply]
The most recent version of the author's plugin in both files in this folder:
https://github.com/ithcy/nicotine-plugins/tree/main/Autobahn

**Reply by iioniis** [reply]
I just put those most recent files into my /Users/[username]/.local/share/nicotine/plugins/autobahn/ directory and it isn't showing up in the plugins menu in Nicotine.  Rebooted Nicotine a couple times and toggled the off/on switch, restarted and nothing.

**Reply by slook** [reply]
That plugin can't load because of the outdated invalid references to `frame`.

If you look in the Log History pane you might see a stack trace listing the lines of code which fail.

**Reply by iioniis** [reply]
Does it have to do with:

self.resolved 

?

**Reply by iioniis** [reply]
Or:

`self.frame`
`self.frame.np`
`self.frame.userbrowse`


I'm guessing  `self.frame`. or just `self`


Please keep in mind I have no idea what I'm talking about.  8]

**Reply by tegitting** [reply]
It sounds like you now got the client to recognise the file, are you now seeing these messages in the debug/information window?

**Reply by iioniis** [reply]
Unfortunately not.  I'm only reading from the file.

I am though looking at the debugger with Leech Detector on and I'm not seeing `self` anything...  only `user`

**Reply by tegitting** [reply]
Which version of Mac do you have, Newer M1/Arm64 or slightly older, I may have figured something out which you can possibly use in the meantime.

**Reply by iioniis** [reply]
I have an Intel Mac - Mid-2015 Pro running Monterey 12.7.3

Would the classification of the plugin: "BasePlugin" have anything to do with it?  Is a BasePlugin not what the built-in plugins are?

**Reply by mathiascode** [reply]
Maybe the plugin files were not downloaded properly? The following steps worked fine for me on macOS:

- Go to https://github.com/ithcy/nicotine-plugins
- Press the green "Code" button -> Download ZIP
- Open the unzipped folder. It will contain an Autobahn folder.
- In Nicotine+, go to Preferences -> Plugins -> Add Plugins. It will open the "plugins" folder.
- Drag the "Autobahn" folder to the "plugins" folder.
- Close and reopen the Preferences dialog in Nicotine+.

The plugin will not work in its current form, of course, but it should at least show up in the list.

**Reply by tegitting** [reply]
Don't ask me how I did this, as I am still trying to fully grasp all this myself, but if you are game, you may try this out. https://github.com/tegitting/nicotine-plus/actions/runs/8027350167/artifacts/1271425018

It should download a ZIP file, inside which is a MAC build. You can see the details of it here : https://github.com/tegitting/nicotine-plus/actions/runs/8027350167/job/21931210249

What it will hopefully install is a Nicotine client that looks just like the one you are running, no extra plugins or anything like that but in this one the built in Leech Detector plugin will behave different than the one you have in that, upon someone attempting to download from you it will auto-browse their shares (you can see the shares). If they aren't sharing whatever you configure in Leech Detector they are auto-banned and sent whatever message you configure in the addon. (as per the addon you are trying to install). 

Note: The number of files is not yet checking Private/Non-private shares so a file is a file, hence it auto-browsing so you can look and see. It's also not checking the number of bytes shared etc like the plugin you are trying to install but it's a first step/attempt.

In summary, this is just regular Nicotine+ 3.3.0 with a (hopefully) modified Leech Detector.
I'd be interested in finding out if it works for you.

**Reply by mathiascode** [reply]
I want to make the process of installing plugins easier at some point, but I don't know when I'll have time for this.

**Reply by tegitting** [reply]
I'm sorry I didnt know Mathias was responding to you. You should try his suggestion first but we know that even if the suggestion works the autobahn addon won't due to the old code references to self.frame but at least it would prove to you the process of custom plugins being recognised by the client.

**Reply by iioniis** [reply]
I will definitely do what mathiascode suggested.  Let me get some food and I'll give it a shot.  Thanks!

**Reply by mathiascode** [reply]
@tegitting Please try renaming your plugin from leech_detector to something else, and see if it's picked up as a custom plugin. Having people stuck on different custom Nicotine+ builds will become a pain to deal with.

**Reply by iioniis** [reply]
@mathiascode , it worked!

**Reply by iioniis** [reply]
so I just got this report, I haver both leech detector and autobahn on:



2024-02-23 20:32:59 Plugin Autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/Users/runner/work/nicotine-plus/nicotine-plus/pynicotine/pluginsystem.py", line 943, in _trigger_event
  File "/Users/iion/.local/share/nicotine/plugins/Autobahn/__init__.py", line 66, in upload_queued_notification
    self.frame.np.userbrowse.browse_user(user)
    ^^^^^^^^^^
2024-02-23 20:32:59 Plugin Autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/Users/runner/work/nicotine-plus/nicotine-plus/pynicotine/pluginsystem.py", line 943, in _trigger_event
  File "/Users/iion/.local/share/nicotine/plugins/Autobahn/__init__.py", line 66, in upload_queued_notification
    self.frame.np.userbrowse.browse_user(user)
    ^^^^^^^^^^
2024-02-23 20:32:59 Plugin Autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/Users/runner/work/nicotine-plus/nicotine-plus/pynicotine/pluginsystem.py", line 943, in _trigger_event
  File "/Users/iion/.local/share/nicotine/plugins/Autobahn/__init__.py", line 66, in upload_queued_notification
    self.frame.np.userbrowse.browse_user(user)
    ^^^^^^^^^^
2024-02-23 20:32:59 Leech Detector: User kzebriquane has no shared files according to the server, requesting shares to verify…
2024-02-23 20:32:59 Plugin Autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/Users/runner/work/nicotine-plus/nicotine-plus/pynicotine/pluginsystem.py", line 943, in _trigger_event
  File "/Users/iion/.local/share/nicotine/plugins/Autobahn/__init__.py", line 115, in user_stats_notification
    total_size_text = self.frame.userbrowse.pages[user].share_size_label.get_text()
                      ^^^^^^^^^^
2024-02-23 20:32:59 Leech Detector: User danaleola is okay, sharing 10570 files in 1383 folders.
2024-02-23 20:32:59 Plugin Autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/Users/runner/work/nicotine-plus/nicotine-plus/pynicotine/pluginsystem.py", line 943, in _trigger_event
  File "/Users/iion/.local/share/nicotine/plugins/Autobahn/__init__.py", line 115, in user_stats_notification
    total_size_text = self.frame.userbrowse.pages[user].share_size_label.get_text()
                      ^^^^^^^^^^
2024-02-23 20:32:59 Leech Detector: Leecher detected, spacedustspace is only sharing 103 files in 1 folders. Going to message leecher after transfer…
2024-02-23 20:32:59 Plugin Autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/Users/runner/work/nicotine-plus/nicotine-plus/pynicotine/pluginsystem.py", line 943, in _trigger_event
  File "/Users/iion/.local/share/nicotine/plugins/Autobahn/__init__.py", line 115, in user_stats_notification
    total_size_text = self.frame.userbrowse.pages[user].share_size_label.get_text()
                      ^^^^^^^^^^
2024-02-23 20:32:59 Leech Detector: Leecher detected, kzebriquane is only sharing 0 files in 1 folders. Going to message leecher after transfer…
2024-02-23 20:32:59 Plugin Autobahn failed with error <class 'AttributeError'>: 'Plugin' object has no attribute 'frame'.
Trace:   File "/Users/runner/work/nicotine-plus/nicotine-plus/pynicotine/pluginsystem.py", line 943, in _trigger_event
  File "/Users/iion/.local/share/nicotine/plugins/Autobahn/__init__.py", line 115, in user_stats_notification
    total_size_text = self.frame.userbrowse.pages[user].share_size_label.get_text()
                      ^^^^^^^^^^

**Reply by iioniis** [reply]
@tegitting I don't know how, but for some reason I missed this.  Do you still have this?

**Reply by tegitting** [reply]
Hi iioniis, I'm not sure what you are referring to. Sorry.

**Comment by tegitting** [comment]
@mathiascode,

I certainly will. I didnt even mean for this to be public I was just experimenting with customising existing code and building for windows. Is it true that a .py file cannot be added to windows Nicotine+ client in the same way it can for MAC + Linux?

I have tried now to push changes and will look to make this a private source to prevent any maintenance issues creeping in near official team.

**Comment by iioniis** [comment]
I was referring to this:

https://github.com/nicotine-plus/nicotine-plus/discussions/2868#discussioncomment-8574260

**Comment by abuckiew** [comment]
Is it possible to either pass in the msg.list and msg.privatelist into the pluginhandler.user_stats_notification, or calculate the sizes and pass them in here? 
https://github.com/nicotine-plus/nicotine-plus/blob/417a9122782905722da14203c29739589c7e3a17/pynicotine/userbrowse.py#L361

This plugin (autobahn) used to calculate the sizes when it had access to the UI object, so giving the plugins access to the sizes (like from the UI's create_folder_tree function) would allow it to keep the same functionality in 3.3+.

**Reply by mathiascode** [reply]
Calculating the sizes and passing them seems reasonable.

**Reply by mathiascode** [reply]
Added in https://github.com/nicotine-plus/nicotine-plus/commit/b3f063ad03b41067796ad8bb8b27177164ec8500.

**Comment by mathiascode** [comment]
This discussion has run its course.
